### PR TITLE
[fix] Add option to wrap root module in auto_wrap

### PR DESCRIPTION
**Pull request by ruanslv**
## What does this PR do?
Fixes # [791](https://github.com/facebookresearch/fairscale/issues/791).

The issue is that the root/outer module sometimes is not wrapped, depending on whether there are enough params left after children are covered by recursive wrapping: https://github.com/facebookresearch/fairscale/blob/main/fairscale/nn/wrap/auto_wrap.py#L299

So I've added a new input param to `default_auto_wrap_policy` to tell us if the current module is the root, and a new flag that allows us to wrap the root regardless of number of unwrapped params (flag is enabled by default).

Adjusted test-cases to show more clearly behavior with and without the new flag when there are not enough params left for root module.

### CODE REVIEW

File: tests/nn/wrap/test_wrap.py
Comment (min-xu-ai) [inline tests/nn/wrap/test_wrap.py:73]: Maybe you have considered this option, but what if we add an option to the default_auto_wrap_policy for this behavior? Adding this option to auto_wrap would affect all policy functions that users may write for auto_wrap? Does that make policy functions harder to write because of this option?

Again, thank you so much for the PR and awesome test cases. My general feeling is that if we can limit options in auto_wrap and try to let the policy function control wrapping, that'd be better. What do you think @ruanslv ?

**Review by min-xu-ai — APPROVED** [review]
This is so cool! Thanks for taking the time to implement this approach! We can merge once all tests are green.

**Comment by min-xu-ai** [comment]
I triggered a rerun of the tests since the failures seems to be due to a single flaky test.

File: pyproject.toml
Comment (min-xu-ai) [inline pyproject.toml:30]: I think you need to undo this change or the precommit check fails.

Comment (ruanslv) [reply pyproject.toml:30]: https://github.com/ruanslv/fairscale/blob/auto_wrap_fix/CONTRIBUTING.md#pre-commit-recommended

These changes came from the pre-commit checks, I'm not managing to do a commit without them. Should I just bypass them? Any idea how to do that?

Comment (min-xu-ai) [reply pyproject.toml:30]: Yeah. Need to run precommit uninstall to bypass it for now.

Comment (anj-s) [reply pyproject.toml:30]: You can also remove the change, run `git commit -m "" --no-verify` to enable pushing to the branch without running a local pre-commit check. There is some mismatch between the local and remove pre-commit checks that i am looking into.

File: fairscale/nn/wrap/auto_wrap.py
Comment (anj-s) [inline fairscale/nn/wrap/auto_wrap.py:21]: Won't this change behavior is users are currently using this policy? Wondering if users will end up doing a double wrap without knowing we've already wrapped the root module.

Comment (min-xu-ai) [reply fairscale/nn/wrap/auto_wrap.py:21]: good catch. It is worth checking. If so, perhaps default it the False? Also, I can make a separate change where we disable double wrapping in FSDP's init function.

Comment (anj-s) [reply fairscale/nn/wrap/auto_wrap.py:21]: Makes sense. Let's make the default behavior as  False for now and once we have the PR that disables double wrapping in FSDP's init function, we can make the default behavior as being True. I was thinking if there might be any repercussions for nested wrapping etc. but I think not wrapping if the module has already been wrapped is a fair check in any condition.

**Comment by min-xu-ai** [comment]
I have a PR #933 that will improve the test stability. Maybe you can merge with main again once that PR is merged?

File: pyproject.toml
Comment (ruanslv) [reply pyproject.toml]: Thanks for the command pointer. I've removed the changes to this file.

**Comment by anj-s** [comment]
@ruanslv Does this need to merge from main before being submitting? Maybe you already did this since I don't see a conflict  above. One last thing is to add an entry to the  [CHANGELOG](https://github.com/facebookresearch/fairscale/blob/main/CHANGELOG.md). Thanks!

**Comment by ruanslv** [comment]
@anj-s I think I'm missing "write access"? Sending screenshot of what I'm seeing attached, I don't see the option to merge. There are also some test failures that seem unrelated to my change, not sure whether they are blocking the merge. Sorry, not super familiar with github UI.
Screenshot 2022-02-15 at 21 16 41

**Comment by anj-s** [comment]
> @anj-s I think I'm missing "write access"? Sending screenshot of what I'm seeing attached, I don't see the option to merge. There are also some test failures that seem unrelated to my change, not sure whether they are blocking the merge. Sorry, not super familiar with github UI. Screenshot 2022-02-15 at 21 16 41


Ah I meant merge from main branch so that there are no conflicts but I don't see any. I am going merge the PR now myself and figure out why you may not have access.

**Comment by ruanslv** [comment]
Oh right, yes I don't think there were any conflicts. Merge worked fine:
```
$ git pull
Already up to date.
ruanrms-mbp:fairscale ruanrms$ git merge auto_wrap_fix
Updating 67bf5bf..8ff6fab
Fast-forward
 fairscale/nn/wrap/auto_wrap.py | 35 +++++++++++++++++++++++++++--------
 tests/nn/wrap/test_wrap.py     | 30 +++++++++++++++++++++++-------
 2 files changed, 50 insertions(+), 15 deletions(-)
```

Thanks for merging the PR. Here are the changelog changes: https://github.com/facebookresearch/fairscale/pull/935

**Pull request merged**
